### PR TITLE
MKLDNN RNN Inference Integration(fp32 LSTM and vRNN with tanh and relu)

### PR DESCRIPTION
**Pull request by lihaofd**
## Description ##
In this PR, it integrated MKLDNN RNN Inference Integration(fp32 lstm and vRNN with tanh and relu)
@pengzhao-intel, @TaoLv , @ciyongch 

### New features ###
- Single layer/Multiple layer and unidirectional/bidirectional inference by mkldnn lstm and vrnn with tanh and relu)

### Unit-test changes ###
- Using existing test case in test_operator.py to check consistency with original RNN Cell implementation.

### Performance ###
We have tested performance of FusedRNN  (USE_MKLDNN = 0 and 1) on our local Skylake-8180 with 1 Sockets and 28 cores. Use MKL as blas lib in this performance test.
Test input size is from DS2 default parameters(seq_length = 300, batch_size = 20, input_size = 800, hidden_size = 800). with MKLDNN commit 57e1203092f63941475ec4088ccd3cf609ed9d7a

Layer=1 bidirectional = False

| API | MKLDNN = 0 (sample/sec) |  MKLDNN = 1  (sample/sec)    |  speedup
| --------           | :-----:                              | :----:                          | :----:            |
| FusedLSTM  | 255                       |  637                           | 2.5x      |
| FusedvRNN with tanh            | 989                      |  1449                  |    1.47x |
| FusedvRNN with relu         | 1296       |  1442     |             1.11x                       | 

Layer=5 bidirectional = True

| API | MKLDNN = 0 (sample/sec) |  MKLDNN = 1  (sample/sec)    |  speedup
| --------           | :-----:                              | :----:                          | :----:            |
| FusedLSTM  | 26                       | 56                           | 2.15x      |
| FusedvRNN with tanh            |  83                     |   157              |    1.9x                  |
| FusedvRNN with relu         | 104             |  152                    |             1.46x                       | 

## Checklist ##
 - [X] Passed code style checking (make lint).
 - [X] All changes have test coverage.
 - [ ] Code is well-documented.

### CODE REVIEW

**Comment by pengzhao-intel** [comment]
@lihaofd We need to upgrade MKL-DNN by a separated PR but you can use this PR for the CI testing.

**Comment by pengzhao-intel** [comment]
FYI @anirudh2290 @szha we are starting the MKLDNN RNN integration :) 
The order of PR is :  inference -> training -> INT8 -> ....

**Comment by anirudh2290** [comment]
Great to hear! Looking forward to this

File: src/operator/nn/mkldnn/mkldnn_rnn_impl.h [hunk 1]
```diff
@@ -0,0 +1,710 @@
...
+  }
+  std::vector<primitive::at> inputs;
+  for (size_t i = 0; i < srcs_cds.size(); i++) {
+    inputs.push_back(srcs[i]);
+  }
+  auto dst_desc = mkldnn::memory::desc(dst_cds, mkldnn_dtype, dst_format);
+  auto concat_pd = concat::primitive_desc(dst_desc, concat_dimension, srcs_pd);
+  MKLDNNStream::Get()->RegisterPrim(concat(concat_pd, inputs, dst));
+  MKLDNNStream::Get()->Submit();
+}
```
Comment (TaoLv) [inline src/operator/nn/mkldnn/mkldnn_rnn_impl.h]: Can we leverage the concat implementation in mkldnn_concat.cc? Do you think the concat primitive here need be cached?

File: src/operator/nn/mkldnn/mkldnn_rnn_impl.h [hunk 2]
```diff
@@ -0,0 +1,710 @@
...
+
+inline size_t GetMKLDNNRNNCacheMemorySize(int L,
```
Comment (TaoLv) [inline src/operator/nn/mkldnn/mkldnn_rnn_impl.h]: Please add comments for each parameter.

File: src/operator/nn/mkldnn/mkldnn_rnn_impl.h [hunk 3]
```diff
@@ -0,0 +1,710 @@
...
+// unidirectional will be done by fused 1 + fused (L - 1) layers or fused L layers(when I = H)
+
+template <typename DType>
+void MKLDNNRNNForwardSingleLayerBi(bool state_outputs,
+                                   const int T,
+                                   const int N,
+                                   const int I,
+                                   const int H,
+                                   DType* x_ptr,
+                                   mkldnn::memory user_src_layer_memory,
```
Comment (TaoLv) [inline src/operator/nn/mkldnn/mkldnn_rnn_impl.h]: Don't know what's this parameter but can we pass by reference?

File: src/operator/nn/mkldnn/mkldnn_rnn_impl.h [hunk 4]
```diff
@@ -0,0 +1,710 @@
...
+  mkldnn::memory::dims weights_layer_tz = {1, 2, I, ngates, H};  //  ldigo
+  mkldnn::memory::dims weights_layer_r_tz = {1, 1, I, ngates, H};  //  ldigo for reorder
+  mkldnn::memory::dims weights_iter_tz = {1, 2, H, ngates, H};  //  ldigo
+  mkldnn::memory::dims weights_iter_r_tz = {1, 1, H, ngates, H};  //  ldigo for reorder
+  mkldnn::memory::dims bias_tz = {1, 2, ngates, H};
+  mkldnn::memory::dims src_iter_tz = {1, 2, nstates, N, H};  //  ldsnc
+  mkldnn::memory::dims dst_iter_tz = {1, 2, nstates, N, H};  //  ldsnc
+
+  std::vector<float> weights_scales(ngates * H);
+  if (!cached) {
```
Comment (TaoLv) [inline src/operator/nn/mkldnn/mkldnn_rnn_impl.h]: What's cached? How is it cached?

File: src/operator/nn/mkldnn/mkldnn_rnn_impl.h [hunk 5]
```diff
@@ -0,0 +1,710 @@
...
+  mkldnn::memory::dims dst_layer_tz = {T, N, 2 * H};
+  mkldnn::memory::dims weights_layer_tz = {1, 2, I, ngates, H};  //  ldigo
+  mkldnn::memory::dims weights_layer_r_tz = {1, 1, I, ngates, H};  //  ldigo for reorder
+  mkldnn::memory::dims weights_iter_tz = {1, 2, H, ngates, H};  //  ldigo
+  mkldnn::memory::dims weights_iter_r_tz = {1, 1, H, ngates, H};  //  ldigo for reorder
+  mkldnn::memory::dims bias_tz = {1, 2, ngates, H};
+  mkldnn::memory::dims src_iter_tz = {1, 2, nstates, N, H};  //  ldsnc
+  mkldnn::memory::dims dst_iter_tz = {1, 2, nstates, N, H};  //  ldsnc
+
+  std::vector<float> weights_scales(ngates * H);
```
Comment (TaoLv) [inline src/operator/nn/mkldnn/mkldnn_rnn_impl.h]: Seems it's not used?

File: src/operator/rnn.cc [hunk 6]
```diff
@@ -138,6 +138,21 @@ static bool RNNType(const nnvm::NodeAttrs& attrs,
...
 }
 
+inline static bool RNNStorageType(const nnvm::NodeAttrs& attrs,
+                                  const int dev_mask,
+                                  DispatchMode* dispatch_mode,
+                                  std::vector<int> *in_attrs,
+                                  std::vector<int> *out_attrs) {
+  DispatchMode wanted_mode = DispatchMode::kFCompute;
+
+  #if MXNET_USE_MKLDNN_RNN
```
Comment (pengzhao-intel) [inline src/operator/rnn.cc]: Why not "USE_MKLDNN"? Is there any conflict?

Comment (lihaofd) [reply src/operator/rnn.cc]: If using USE_MKLDNN here, several GPU case in ci will fail

File: src/operator/nn/mkldnn/mkldnn_rnn_impl.h (hunk 1 above)
Comment (lihaofd) [reply src/operator/nn/mkldnn/mkldnn_rnn_impl.h]: There are too many data segments with different size, dim, count and concat_dim etc. It will make mkldnn cache be much more complicated but will not benefit too much on perf

File: src/operator/nn/mkldnn/mkldnn_rnn_impl.h (hunk 4 above)
Comment (lihaofd) [reply src/operator/nn/mkldnn/mkldnn_rnn_impl.h]: At the first time, it will did data preparation like wx, wh concat/reorder etc for multiple layers unidirectional or bidirectional ways and saved them into mkldnn cached memory. From next time call, these data be used directly.

File: src/operator/nn/mkldnn/mkldnn_rnn_impl.h (hunk 5 above)
Comment (lihaofd) [reply src/operator/nn/mkldnn/mkldnn_rnn_impl.h]: fixed

File: src/operator/nn/mkldnn/mkldnn_rnn_impl.h (hunk 3 above)
Comment (lihaofd) [reply src/operator/nn/mkldnn/mkldnn_rnn_impl.h]: fixed

File: src/operator/nn/mkldnn/mkldnn_rnn_impl.h (hunk 2 above)
Comment (lihaofd) [reply src/operator/nn/mkldnn/mkldnn_rnn_impl.h]: fixed

File: src/operator/rnn.cc (hunk 6 above)
Comment (pengzhao-intel) [reply src/operator/rnn.cc]: What's the error message? Seems we need to fix this issue.

Comment (lihaofd) [reply src/operator/rnn.cc]: many error messages like src/operator/./rnn-inl.h(1328): error: argument of type "mshadow::Stream<mshadow::cpu> *" is incompatible with parameter of type "mshadow::Stream<mshadow::gpu> *"

**Comment by anirudhacharya** [comment]
@mxnet-label-bot update [pr-work-in-progress]

File: src/operator/nn/mkldnn/mkldnn_rnn_impl.h [hunk 7]
```diff
@@ -0,0 +1,739 @@
...
+                                   std::vector<mkldnn::memory> *bias_memory,
+                                   std::vector<mkldnn::memory> *y_memory,
+                                   std::vector<mkldnn::memory> *hcy_memory,
+                                   std::vector<primitive> *rnn_forward_prim,
+                                   int layer_index,
+                                   bool *has_cache,
+                                   int lvalue,
+                                   int dtype,
+                                   bool is_train,
+                                   int mode) {
```
Comment (pengzhao-intel) [inline src/operator/nn/mkldnn/mkldnn_rnn_impl.h]: could we build a structure or something to simplify the parameters?

File: src/operator/nn/mkldnn/mkldnn_rnn_impl.h [hunk 8]
```diff
@@ -0,0 +1,739 @@
...
+  const int b_size = 2 * H * ngates * D;
+  const int cell_size = N * H * D;
+  //  First layer
+  int w_size = (I + H) * H * ngates * D;
+  auto cpu_engine = CpuEngine::Get()->get_engine();
+  auto null_memory_ = null_memory(cpu_engine);
+  DType* tmpNull = NULL;
+  // when D = 1 and I == H, L layers can be fused together
+  if (D == 1 && I == H) {
+    MKLDNNRNNForwardUnidi(state_outputs, L, T, N, I, H, x_ptr, &null_memory_,
```
Comment (pengzhao-intel) [inline src/operator/nn/mkldnn/mkldnn_rnn_impl.h]: Too many sub-function for different cases and I'm afraid it's really error-prone.

**Comment by pengzhao-intel** [comment]
@zixuanweeei please refresh the performance data with the latest master code since the MKL-DNN has been upgraded to 0.19.

**Comment by zixuanweeei** [comment]
I re-evaluated the performance with MKLDNN commit 41bee20d7eb4a67feeeeb8d597b3598994eb1959 on local Skylake-8180 with 1 Sockets and 28 cores. Here is the result.

### Layer=1 bidirectional = False
|API|MKLDNN = 0 (sample/sec)|MKLDNN = 1(sample/sec)|speedup|
|---|-----------------------|----------------------|-------|
|FusedLSTM|289              |682                   |2.36x  |
|FusedvRNN with tanh|733    |1408                  |1.43x  |
|FusedvRNN with relu|1074   |1375                  |1.28x  |

### Layer=5 bidirectional = True
|API|MKLDNN = 0 (sample/sec)|MKLDNN = 1(sample/sec)|speedup|
|---|-----------------------|----------------------|-------|
|FusedLSTM|28               |65                    |2.32x  |
|FusedvRNN with tanh|70     |149                   |2.13x  |
|FusedvRNN with relu|94     |147                   |1.56x  |

**Comment by szha** [comment]
@lihaofd rebase?

**Comment by pengzhao-intel** [comment]
@szha WIP for this :)

**Comment by pengzhao-intel** [comment]
> I re-evaluated the performance with MKLDNN commit 41bee20d7eb4a67feeeeb8d597b3598994eb1959 on local Skylake-8180 with 1 Sockets and 28 cores. Here is the result.
> 
> ### Layer=1 bidirectional = False
> API	MKLDNN = 0 (sample/sec)	MKLDNN = 1(sample/sec)	speedup
> FusedLSTM	289	682	2.36x
> FusedvRNN with tanh	733	1408	1.43x
> FusedvRNN with relu	1074	1375	1.28x
> ### Layer=5 bidirectional = True
> API	MKLDNN = 0 (sample/sec)	MKLDNN = 1(sample/sec)	speedup
> FusedLSTM	28	65	2.32x
> FusedvRNN with tanh	70	149	2.13x
> FusedvRNN with relu	94	147	1.56x

Looks good!
Could you summarize all combinations covered by this PR and test it?

File: src/operator/rnn-inl.h [hunk 9]
```diff
@@ -829,22 +850,23 @@ class RNNOp {
...
+        }
+        if (!temp_init_space_) {
+          temp_cpu_space_ = Storage::Get()->Alloc
+              (work_cpu_space_size * sizeof(DType), Context::CPU());
+          temp_cpu_space_size_ = work_cpu_space_size;
+          temp_init_space_ = true;
+        }
+        DType* work_cpu_space = static_cast<DType*>(temp_cpu_space_.dptr);
+
         const size_t r_size = GetRNNReserveSpaceSize(param_.num_layers, direction,
```
Comment (pengzhao-intel) [inline src/operator/rnn-inl.h]: Why move this part from L832 to here?

File: src/operator/rnn-inl.h [hunk 10]
```diff
@@ -880,6 +902,89 @@ class RNNOp {
                                   param_.p,
                                   param_.mode);
       } else {
+        #if MXNET_USE_MKLDNN == 1 && !defined(__CUDACC__)
+        if (param_.mode != rnn_enum::kGru) {
```
Comment (pengzhao-intel) [inline src/operator/rnn-inl.h]: Add `TODO` for GRU

File: src/operator/nn/mkldnn/mkldnn_rnn_impl.h [hunk 11]
```diff
@@ -0,0 +1,739 @@
...
+  DType* bias_reset = bias;
+  DType* bias_update = bias + H;
+  #pragma omp parallel for num_threads(omp_threads)
+  for (int i = 0; i < H; i++) {
+    DType tmp = bias_update[i];
+    bias_update[i] = bias_reset[i];
+    bias_reset[i] = tmp;
+  }
+}
+// since there is different sematics of MKLDNN's Fused RNN and Mxnet FusedRNN,
```
Comment (pengzhao-intel) [inline src/operator/nn/mkldnn/mkldnn_rnn_impl.h]: Mxnet to MXNet

**Comment by zixuanweeei** [comment]
Here is the test for all combinations covered by this PR. We evaluated the performance with MKLDNN commit 41bee20d7eb4a67feeeeb8d597b3598994eb1959 on local Skylake-6148 with 1 Sockets and 20 cores.

### Layer=1 bidirectional = False
|API|MKLDNN = 0 (sample/sec)|MKLDNN = 1(sample/sec)|speedup|
|---|-----------------------|----------------------|-------|
|FusedLSTM|290              |659                   |2.27x  |
|FusedvRNN with tanh|978    |1367                  |1.40x  |
|FusedvRNN with relu|1297   |1341                  |1.03x  |

### Layer=1 bidirectional = True
|API|MKLDNN = 0 (sample/sec)|MKLDNN = 1(sample/sec)|speedup|
|---|-----------------------|----------------------|-------|
|FusedLSTM|155              |340                   |2.19x  |
|FusedvRNN with tanh|479    |703                   |1.47x  |
|FusedvRNN with relu|627    |684                   |1.09x  |

### Layer=5 bidirectional = False
|API|MKLDNN = 0 (sample/sec)|MKLDNN = 1(sample/sec)|speedup|
|---|-----------------------|----------------------|-------|
|FusedLSTM|64               |144                   |2.25x  |
|FusedvRNN with tanh|215    |327                   |1.52x  |
|FusedvRNN with relu|286    |320                   |1.12x  |

### Layer=5 bidirectional = True
|API|MKLDNN = 0 (sample/sec)|MKLDNN = 1(sample/sec)|speedup|
|---|-----------------------|----------------------|-------|
|FusedLSTM|29               |61                    |2.10x  |
|FusedvRNN with tanh|91     |142                   |1.56x  |
|FusedvRNN with relu|119    |138                   |1.16x  |

**Comment by pengzhao-intel** [comment]
@TaoLv please take a review.
@DickJC123 could you help to take a review if our change is fine for GPU?

**Comment by pengzhao-intel** [comment]
The PR is almost done and we're waiting for the local test.
@TaoLv will update the results soon.

File: src/operator/rnn-inl.h [hunk 12]
```diff
@@ -880,23 +902,78 @@ class RNNOp {
...
-                                   x.dptr_,
-                                   hx.dptr_,
-                                   cx_ptr,
-                                   w.dptr_,
-                                   b_ptr,
-                                   y.dptr_,
-                                   hy_ptr,
-                                   cy_ptr,
-                                   param_.mode);
+        #if MXNET_USE_MKLDNN == 1 && !defined(__CUDACC__)
```
Comment (pengzhao-intel) [inline src/operator/rnn-inl.h]: Is this __CUDACC__ forgot or intend to leave?

Comment (TaoLv) [reply src/operator/rnn-inl.h]: Yes, I think it can be removed also. @zixuanweeei

File: src/operator/rnn-inl.h [hunk 13]
```diff
@@ -880,23 +902,78 @@ class RNNOp {
...
+        if (dmlc::GetEnv("MXNET_USE_MKLDNN_RNN", 1) && param_.mode != rnn_enum::kGru) {
```
Comment (TaoLv) [inline src/operator/rnn-inl.h]: @szha Please review. We add a environmental variable here. Once it's set to 0, RNN operator will fall back to the original version on CPU. Otherwise, MKL-DNN RNN primitive will be invoked.

File: src/operator/rnn-inl.h (hunk 12 above)
Comment (zixuanweeei) [reply src/operator/rnn-inl.h]: It has been modified. Thanks.

File: src/operator/rnn-inl.h [hunk 14]
```diff
@@ -880,23 +902,78 @@ class RNNOp {
...
+                                     x.dptr_,
+                                     hx.dptr_,
+                                     cx_ptr,
+                                     w.dptr_,
+                                     b_ptr,
+                                     y.dptr_,
+                                     hy_ptr,
+                                     cy_ptr,
+                                     param_.mode);
+        #if MXNET_USE_MKLDNN == 1 && !defined(__CUDACC__)
```
Comment (pengzhao-intel) [inline src/operator/rnn-inl.h]: @zixuanweeei  And this one :(

**Comment by zixuanweeei** [comment]
We also test the performances over GPU of our PR and the master. Here is the the result. The relative DIFF is calculated by (Our_PR - MASTER) / MASTER. In summary, Our modifications have no significant damage to the performance over GPU.

### Layer = 1, bidirectional = False
| API | Our PR (sample/sec) | MASTER (sample/sec) | DIFF (sample/sec) | Relative DIFF |
|-----|---------------------|---------------------|-------------------|---------------|
|FusedLSTM	|1038	|1058	|-20	|-1.89%|
|FusedvRNN with tanh	|1961	|1884	|77	|4.09%
|FusedvRNN with relu	|1926	|1939	|-13	|-0.67%|

### Layer = 1, bidirectional = True
| API | Our PR (sample/sec) | MASTER (sample/sec) | DIFF (sample/sec) | Relative DIFF |
|-----|---------------------|---------------------|-------------------|---------------|
|FusedLSTM	|683	|694	|-11	|-1.59%|
|FusedvRNN with tanh	|1221	|1190	|31	|2.61%|
|FusedvRNN with relu	|1212	|1209	|3	|0.25%|

### Layer = 5, bidirectional = False
| API | Our PR (sample/sec) | MASTER (sample/sec) | DIFF (sample/sec) | Relative DIFF |
|-----|---------------------|---------------------|-------------------|---------------|
|FusedLSTM	|322	|320	|2	|0.63%|
|FusedvRNN with tanh	|676	|649	|27	|4.16%|
|FusedvRNN with relu	|670	|637	|33	|5.18%|

### Layer = 5, bidirectional = True
| API | Our PR (sample/sec) | MASTER (sample/sec) | DIFF (sample/sec) | Relative DIFF |
|-----|---------------------|---------------------|-------------------|---------------|
|FusedLSTM	|110	|109	|1	|0.92%|
|FusedvRNN with tanh	|210	|209	|1	|0.48%|
|FusedvRNN with relu	|212	|210	|2	|0.95%|

**Comment by DickJC123** [comment]
I think it prudent to resolve the GPU platform issues with rnn-inl.h introduced by commit 1c49e40fd before finally accepting this PR [see https://github.com/apache/incubator-mxnet/issues/15034].  Besides introducing test failures of test_rnntanh_bidirectional on P40 GPUs, I have noticed that the codebase no longer compiles against cuDNN versions < 7.0.  I intend to submit a PR to resolve both these issues within 24 hours, probably less.

**Comment by pengzhao-intel** [comment]
The PR is ready to be merged. 

@szha @DickJC123 do we need to wait for #15056 ?

**Comment by pengzhao-intel** [comment]
Merging this one first since the #15056 WIP.
